scalyr_agent/copying_manager/copying_manager.py

```diff
@@ -72,6 +74,62 @@
 WORKER_SESSION_PROCESS_MONITOR_ID_PREFIX = "agent_worker_session_"
 
 
+class DynamicWorkers(object):
+    """
+    This class is responsible for managing the dynamic workers.
+    """
+
+    def __init__(self):
+        self.__workers_by_api_key = {}
+        self.__workers_by_id = {}
+        self.__last_id = 0
+        self.__create_worker_lock = threading.Lock()
+
+    def get_worker(self, worker_id):
+        # type: (str) -> Optional[CopyingManagerWorker]
+        """
+        Get worker by id.
+        """
+        return self.__workers_by_id[worker_id]
+
+    def get_or_create_worker(self, api_key, global_config):
+        # type: (str, Configuration) -> CopyingManagerWorker
+        """
+        Add worker.
+        """
+        if not api_key in self.__workers_by_api_key:
+            with self.__create_worker_lock:
+                if not api_key in self.__workers_by_api_key:
+                    worker = self.__create_worker(api_key, global_config)
+                    self.__workers_by_api_key[api_key] = worker
+                    self.__workers_by_id[worker.worker_id] = worker
+                    worker.start_sessions_and_block_until_copying()
+
+        return self.__workers_by_api_key[api_key]
+
+    def values(self):
+        return self.__workers_by_id.values()
+
+    def items(self):
+        return self.__workers_by_id.items()
```

Comment:
With the GIL I think this is ok but as I understand the GIL only applies to Cython-based implementations.  To be on the safe side it might be a good idea to protect these accesses with the create_worker_lock (probably should rename it to worker_lock as well)

Reply:
Good idea.